File: backend/app/schemas/warehouse_transfers.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.schemas.common import OccurredAt
from app.schemas.units import MaterialUnitOut


class WarehouseTransferLineOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    unit: MaterialUnitOut
    added_at: datetime
    received_at: datetime | None


class WarehouseTransferOut(BaseModel):
    id: int
    from_warehouse_id: int
    from_warehouse_name: str
    to_warehouse_id: int
    to_warehouse_name: str
    status: str
    note: str | None
    created_at: datetime
    shipped_at: datetime | None
    received_at: datetime | None
    lines: list[WarehouseTransferLineOut]


class AddUnitToTransferRequest(BaseModel):
    unit_id: int
    to_warehouse_id: int
    note: str | None = None
    occurred_at: OccurredAt = None


class ReceiveTransferLineRequest(BaseModel):
    occurred_at: OccurredAt = None

File: frontend/src/api/warehouseTransfers.ts
import { apiClient } from "./client";
import type { MaterialUnit } from "./units";

export interface WarehouseTransferLine {
  id: number;
  unit: MaterialUnit;
  added_at: string;
  received_at: string | null;
}

export interface WarehouseTransfer {
  id: number;
  from_warehouse_id: number;
  from_warehouse_name: string;
  to_warehouse_id: number;
  to_warehouse_name: string;
  status: "sobiraetsya" | "otpravleno" | "prinyato";
  note: string | null;
  created_at: string;
  shipped_at: string | null;
  received_at: string | null;
  lines: WarehouseTransferLine[];
}

export async function listWarehouseTransfers(params?: {
  status_filter?: string;
  to_warehouse_id?: number;
}): Promise<WarehouseTransfer[]> {
  const { data } = await apiClient.get<WarehouseTransfer[]>("/warehouse-transfers", { params });
  return data;
}

export async function addUnitToTransfer(payload: {
  unit_id: number;
  to_warehouse_id: number;
  note?: string;
  occurred_at?: string;
}): Promise<WarehouseTransfer> {
  const { data } = await apiClient.post<WarehouseTransfer>("/warehouse-transfers/add-unit", payload);
  return data;
}

export async function removeTransferLine(transferId: number, lineId: number): Promise<WarehouseTransfer> {
  const { data } = await apiClient.post<WarehouseTransfer>(`/warehouse-transfers/${transferId}/lines/${lineId}/remove`);
  return data;
}

export async function shipTransfer(transferId: number): Promise<WarehouseTransfer> {
  const { data } = await apiClient.post<WarehouseTransfer>(`/warehouse-transfers/${transferId}/ship`);
  return data;
}

export async function receiveTransferLine(transferId: number, lineId: number, occurredAt?: string): Promise<WarehouseTransfer> {
  const { data } = await apiClient.post<WarehouseTransfer>(`/warehouse-transfers/${transferId}/lines/${lineId}/receive`, {
    occurred_at: occurredAt,
  });
  return data;
}

export async function receiveAllTransferLines(transferId: number, occurredAt?: string): Promise<WarehouseTransfer> {
  const { data } = await apiClient.post<WarehouseTransfer>(`/warehouse-transfers/${transferId}/receive-all`, {
    occurred_at: occurredAt,
  });
  return data;
}
